File: service/src/main/java/com/boxfox/cross/service/PriceService.java
package com.boxfox.cross.service;

import com.boxfox.cross.common.vertx.service.AbstractService;
import com.mashape.unirest.http.Unirest;
import com.mashape.unirest.http.exceptions.UnirestException;
import com.boxfox.cross.common.vertx.service.Service;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import org.json.JSONArray;
import org.json.JSONObject;

public class PriceService extends AbstractService {
    private static final String COINMARKET_CAP_URL = "https://api.coinmarketcap.com/v2/ticker/";
    private Map<String, Integer> coinIdMap;

    @Service
    LocaleService localeService; //@TODO Service inject to Service

    public PriceService() {
        this.coinIdMap = new HashMap<>();
        try {
            JSONArray array = Unirest.get(COINMARKET_CAP_URL + "listings/").asJson().getBody().getObject().getJSONArray("data");
            array.forEach(obj -> {
                JSONObject coin = (JSONObject) obj;
                String symbol = coin.getString("symbol");
                int id = coin.getInt("id");
                coinIdMap.put(symbol, id);
            });
        } catch (UnirestException e) {
            e.printStackTrace();
        }
    }

    public double getPrice(String symbol, double balance) {
        return getPrice(symbol, Locale.KOREA.toString(), balance);
    }

    public double getPrice(String symbol, String locale, double balance) {
        double price = getPrice(symbol, locale);
        if (price > 0) {
            return price * balance; //@TODO dobule calculate improve
        } else {
            return -1;
        }
    }

    public double getPrice(String symbol) {
        return getPrice(symbol, Locale.KOREA.toString());
    }

    public double getPrice(String symbol, String locale) {
        String moneySymbol = localeService.getLocaleMoneySymbol(locale);
        symbol = symbol.toUpperCase();
        if (coinIdMap.containsKey(symbol)) {
            try {
                int id = coinIdMap.get(symbol);
                String url = String.format("%s%s/?convert=%s", COINMARKET_CAP_URL, id, moneySymbol);
                JSONObject obj = Unirest
                        .get(url).asJson()
                        .getBody().getObject();
                JSONObject krw = obj.getJSONObject("data").getJSONObject("quotes").getJSONObject(moneySymbol);
                return krw.getDouble("price");
            } catch (UnirestException e) {
                e.printStackTrace();
            }
        }
        return -1;
    }
}

File: service/src/main/java/com/boxfox/cross/service/price/PriceService.kt
package com.boxfox.cross.service.price

import com.boxfox.cross.service.JooqReactiveService
import com.boxfox.cross.service.LocaleService
import com.boxfox.cross.service.wallet.WalletService
import com.boxfox.cross.service.wallet.WalletServiceImpl
import com.boxfox.vertx.service.Service
import io.reactivex.Single
import java.util.*

class PriceService(private val impl: PriceServiceImpl = PriceServiceImpl(),
                   private val walletImpl: WalletServiceImpl = WalletServiceImpl()) : JooqReactiveService() {

    @Service
    internal lateinit var localeService: LocaleService

    @Service
    lateinit var walletService: WalletService

    fun getTotalPrice(uid: String, locale: String): Single<Double> {
        return createSingle { impl.getTotalPrice(it, walletImpl.getWalletList(it, uid), localeService.getLocaleMoneySymbol(locale)) }
    }

    fun getWalletPrice(address: String, locale: String): Single<Double> {
        return createSingle {
            val wallet = walletImpl.getWallet(it, address)
            impl.getPrice(it, wallet.coinSymbol, localeService.getLocaleMoneySymbol(locale), wallet.balance)

        }
    }

    fun getPrice(symbol: String, balance: Double): Single<Double> {
        return createSingle { impl.getPrice(it, symbol, Locale.KOREA.toString(), balance) }
    }

    fun getPrice(symbol: String, locale: String, balance: Double): Single<Double> {
        return createSingle { impl.getPrice(it, symbol, localeService.getLocaleMoneySymbol(locale), balance) }
    }

    @JvmOverloads
    fun getPrice(symbol: String, locale: String = Locale.KOREA.toString()): Single<Double> {
        return impl.getPrice(symbol, locale)
    }
}
